### ISSUE TITLE
Fix: Allow sample products to be excluded from task completion criteria

### ISSUE THREAD

**Issue by rjchow** [woothemes/woocommerce#36195]
**Issue Description:**
pbIJXs-34V-p2#comment-5797

There's an issue with the task list completion criteria for the "Add Products" task, as it is currently triggering automatically it is detected that any products have been published. This is triggered when [wp_count_posts](https://github.com/woocommerce/woocommerce/blob/e9c4953ab36d738f1f9fd5a316b1b94cc78d9171/plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Products.php#L164) returns a non-zero number for `product` posts.

This also triggers when sample products are automatically added and published by themes or plugins etc, without any of the users' involvement. 

This would cause inaccurate analytics about how many users are successfully completing store setup as it is one of the criteria for that.

The proposed solution is to add hooks to the `has_products()` function that would allow adding and removing filters on `wp_count_posts`, so that automatically added sample products can be excluded from the criteria.

**Comment by ilyasfoo** [comment]
Reference to suggestion by @daledupreez to add hooks: pbIJXs-34V-p2#comment-5797

> Yes, my thinking was that we could remove any products that we added from our code (based on a post_meta field) and which haven’t been updated by the user.

**Comment by daledupreez** [comment]
Might it also be possible to set a custom `post_meta` flag that indicates a product is a sample product?

**Comment by ilyasfoo** [comment]
Yes, I think that was your prior suggestion as well? Unless I'm missing something 🤔

**Comment by psealock** [comment]
>Might it also be possible to set a custom post_meta flag that indicates a product is a sample product?

Based on post meta how can we differentiate sample products when plugins we don't control can add products? Meanwhile, merchants can import either sample products or legitimate products. How can we tell the difference?

For the purposes of task completion, does it even matter if the product was imported/auto-added or was legitimately added by the merchant via the UI? I'm not so sure it matters from the merchant perspective.

When there are products and the merchant has not used the UI to add one, we could add a UI element for merchants to indicate if they are sample products or not. But this feels like we're asking the question to suit our interests instead of the merchant, so I hesitate to advocate for this.

**Comment by ilyasfoo** [comment]
Hey, @psealock! Thanks for raising this. 

> Based on post meta how can we differentiate sample products when plugins we don't control can add products? Meanwhile, merchants can import either sample products or legitimate products. How can we tell the difference?

You're right, we can't for the cases you've highlighted. 

> For the purposes of task completion, does it even matter if the product was imported/auto-added or was legitimately added by the merchant via the UI? I'm not so sure it matters from the merchant perspective.

I believe this issue originated from a concern product task automatically completed in Woo Express straight after site is set up since the default theme (Tsubaki) added sample products. From the merchant's perspective, I think may feel weird since they have not done anything yet.

That has been addressed in https://github.com/Automattic/wc-calypso-bridge/issues/955 for Woo Express/Headstart themes.

> When there are products and the merchant has not used the UI to add one, we could add a UI element for merchants to indicate if they are sample products or not. But this feels like we're asking the question to suit our interests instead of the merchant, so I hesitate to advocate for this.

I kind agree with you on this, I think the remaining concern with Woo core is regarding tracks? pbIJXs-34V-p2#comment-5784 

cc @SamratBiswas1 do you think we should care if the product task is marked as completed even if the product is added by plugins/themes (excluding Woo Express default theme)? If not, I think we can close this issue.

**Comment by adrianduffell** [comment]
> cc @SamratBiswas1 do you think we should care if the product task is marked as completed even if the product is added by plugins/themes (excluding Woo Express default theme)?

I can probably answer this one. I'd say yes we should care, because merchants won't be able to progress to launching their store with only sample products. I suggest we need to add a first class [products property](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-properties) (e.g. `is_sample`) to allow the exclusion of sample products from the count. 

Another possible use-case later might be to delete sample products in the launch your store task once we have these pieces in place 🤔

**Comment by ilyasfoo** [comment]
> I can probably answer this one. I'd say yes we should care, because merchants won't be able to progress to launching their store with only sample products. I suggest we need to add a first class [products property](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-properties) (e.g. is_sample) to allow the exclusion of sample products from the count.

Sounds good 👍

**Comment by timmy5685** [comment]
Hi @rjchow - coming back to this during a triage pass, and it's been resolved — just never closed.

The completion criteria no longer counts raw product posts. `Products.php::has_products()` now counts a product toward the task only if it's published and either the merchant has edited it (`_edit_last` meta) or it isn't a Headstart/sample product (`_headstart_post` meta). So auto-added sample products are excluded — exactly what this asked for. They no longer auto-complete the task or skew the setup-completion analytics.

It landed in two steps:

- #44890 — "[CYS] Mark the products task as completed only if products were user created" — added the `_headstart_post` exclusion
- #56389 — "Fix: Imported CSV products not completing product task" — refined it so imported products still count

The shipped approach keys off the existing `_headstart_post` marker rather than the first-class `is_sample` property discussed later in the thread — which fits, since that's how the Headstart/Tsubaki sample products behind the original Woo Express report are tagged. (The Woo Express side was also handled separately in Automattic/wc-calypso-bridge#955.)

One accepted limit, per @psealock's point above: this only excludes products tagged via Headstart — an arbitrary plugin adding unmarked products would still count. But those are indistinguishable from real products anyway, so it's a reasonable boundary rather than a gap.

Closing as resolved.
